apps/devhub/utils.py

```diff
@@ -84,3 +97,203 @@ def escape_validation(validation):
+
+        # When the validation jobs complete, pass the results to the
+        # appropriate annotate/save task for the object type.
+        self.task = chain(validate, save.subtask([file_.pk]))
+
+    @staticmethod
+    def validate_file(file):
+        """Return a subtask to validate a File instance."""
+        return tasks.validate_file.subtask([file.pk])
+
+    @staticmethod
+    def validate_upload(upload, is_listed):
+        """Return a subtask to validate a FileUpload instance."""
+        assert not upload.validation
+
+        return tasks.validate_upload.subtask([upload.pk, is_listed])
+
+    def find_previous_version(self, version):
+        """Find a the most recent previous version of this add-on, prior to
+        `version`, that we can use to compare validation results."""
+
+        if not self.addon:
+            return
+
+        version = Version(version)
+        statuses = amo.STATUS_PUBLIC, amo.STATUS_LITE
+
+        # Find any version which has been approved, or any beta
+        # version which has been signed.
+        files = (File.objects.filter(Q(version__addon=self.addon) &
+                                     (Q(status__in=statuses) |
+                                      Q(status=amo.STATUS_BETA,
+                                        is_signed=True)))
+                 .order_by('-id'))
+
+        if self.file:
+            # Add some extra filters if we're validating a File instance,
+            # to try to get the closes possible match.
+            files = (files.exclude(pk=self.file.pk)
+                     # Files which are not for the same platform, but have
+                     # other files in the same version which are.
+                     .exclude(~Q(platform=self.file.platform) &
+                              Q(version__files__platform=self.file.platform))
+                     # Files which are not for the either the same platform or
+                     # for all platforms, but have other versions in the same
+                     # version which are.
+                     .exclude(~Q(platform__in=(self.file.platform,
+                                               amo.PLATFORM_ALL.id)) &
+                              Q(version__files__platform=amo.PLATFORM_ALL.id)))
+
+        for file in files:
+            # Only accept versions which come before the one we're validating.
+            if Version(file.version.version) < version:
+                return file
+
+
+class ValidationComparator(object):
+    """Compares the validation results from an older version with a version
+    currently being checked, and annotates results based on which messages
+    may be ignored for the purposes of automated signing."""
+
+    def __init__(self, validation):
+        self.validation = validation
+
+        self.messages = {self.message_key(msg): msg
+                         for msg in validation['messages']
+                         if 'context' in msg}
+
+    @classmethod
```

Comment:
(To be clear, I made these class methods because I'm planning to use this logic to implement the interface that allows editors to flag messages as ignorable/not ignorable. But static methods work as well as class methods for that.)